core/src/repair_weighted_traversal.rs

```diff
@@ -69,10 +69,119 @@ impl<'a> Iterator for RepairWeightTraversal<'a> {
     }
 }
 
+struct GenericTraversal<'a> {
+    tree: &'a HeaviestSubtreeForkChoice,
+    pending: Vec<Slot>,
+}
+
+impl<'a> GenericTraversal<'a> {
+    pub fn new(tree: &'a HeaviestSubtreeForkChoice) -> Self {
+        Self {
+            tree,
+            pending: vec![tree.root().0],
+        }
+    }
+}
+
+impl<'a> Iterator for GenericTraversal<'a> {
+    type Item = Slot;
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.pending.pop();
+        if let Some(slot) = next {
+            let children: Vec<_> = self
+                .tree
+                .children(&(slot, Hash::default()))
+                .unwrap()
+                .iter()
+                .map(|(child_slot, _)| *child_slot)
+                .collect();
+            self.pending.extend(children);
+        }
+        next
+    }
+}
+
+pub fn get_unknown_last_index(
+    tree: &HeaviestSubtreeForkChoice,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    processed_slots: &mut HashSet<Slot>,
+    limit: usize,
+) -> Vec<ShredRepairType> {
+    let mut repairs = Vec::new();
+    let iter = GenericTraversal::new(tree);
+    for slot in iter {
+        if repairs.len() >= limit {
+            break;
+        }
+        if processed_slots.contains(&slot) {
+            continue;
+        }
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.known_last_index().is_none() {
+                repairs.push(ShredRepairType::HighestShred(slot, slot_meta.received));
+                processed_slots.insert(slot);
+            }
+        }
+    }
+    repairs
+}
+
+pub fn get_closest_completion(
+    tree: &HeaviestSubtreeForkChoice,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    processed_slots: &mut HashSet<Slot>,
+    limit: usize,
+) -> Vec<ShredRepairType> {
+    let mut v: Vec<(Slot, u64)> = Vec::default();
+    let iter = GenericTraversal::new(tree);
+    for slot in iter {
+        if processed_slots.contains(&slot) {
+            continue;
+        }
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.is_full() {
+                continue;
+            }
+            if let Some(last_index) = slot_meta.known_last_index() {
+                let dist = last_index - slot_meta.consumed;
```

Comment:
I think the distance might be better to get the total number of shreds in the slot rather than `consumed`, i.e. read the `Index.data.index.len()` from `Blockstore::get_index()`. 

This is because consumed is the index of the last consecutive set of shreds in the slot, so it will return 0, even if you had every shred in the slot other than the first one

Reply:
good call, updated